tensorboard/webapp/plugins/npmi/views/annotations_list/annotation/annotation_component.ts

```diff
@@ -0,0 +1,662 @@
+  @Input() hiddenAnnotations!: string[];
+  @Input() activeMetrics!: string[];
+  @Input() showCounts!: boolean;
+  @Input() annotation!: string;
+  @ViewChild('chart', {static: true, read: ElementRef})
+  private readonly annotationContainer!: ElementRef<HTMLDivElement>;
+  get height(): number {
+    return this.runs.length * 30;
+  }
+  get width(): number {
+    return parseInt(
+      d3.select(this.annotationContainer.nativeElement).style('width'),
+      10
+    );
+  }
+  private readonly margin = {top: 0, right: 0, bottom: 0, left: 100};
+  get chartWidth(): number {
+    return this.width - this.margin.left - this.margin.right;
+  }
+  get chartHeight(): number {
+    return this.height - this.margin.top - this.margin.bottom;
+  }
+  private textClass = 'default-text';
+  get runs(): string[] {
+    const runs = new Set<string>();
+    this.data.forEach((element) => runs.add(element.run));
+    return [...runs];
+  }
+  // Drawing containers
+  private svg!: d3.Selection<
+    SVGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private mainContainer!: d3.Selection<
+    SVGGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private barsGroup!: d3.Selection<
+    SVGGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private countDotsGroup!: d3.Selection<
+    SVGGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private textsGroup!: d3.Selection<
+    SVGGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private countTextsGroup!: d3.Selection<
+    SVGGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private runHintGroup!: d3.Selection<
+    SVGGElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  // Scales
+  private xScale!: d3.ScalePoint<string>;
+  private yScale!: d3.ScalePoint<string>;
+  private sizeScale!: d3.ScaleLinear<number, number>;
+  private countSizeScale!: d3.ScaleLinear<number, number>;
+
+  private runClipPath!: d3.Selection<
+    SVGRectElement,
+    unknown,
+    HTMLElement | null,
+    undefined
+  >;
+  private rgbColors = ['240, 120, 80', '46, 119, 182', '190, 64, 36'];
+
+  ngAfterViewInit(): void {
+    this.svg = d3.select(this.annotationContainer.nativeElement).select('svg');
+    this.xScale = d3.scalePoint<string>().padding(0);
+    this.yScale = d3.scalePoint<string>().padding(0);
+    this.sizeScale = d3.scaleLinear().domain([0.0, 1.0]);
+    this.countSizeScale = d3.scaleLinear().range([2, 10]);
+    this.mainContainer = this.svg
+      .append('g')
+      .attr('transform', `translate(${this.margin.left}, ${this.margin.top})`);
+    this.barsGroup = this.mainContainer.append('g');
+    this.countDotsGroup = this.mainContainer.append('g');
+    this.textsGroup = this.mainContainer.append('g');
+    this.countTextsGroup = this.mainContainer.append('g');
+    this.runHintGroup = this.svg.append('g');
+    this.runClipPath = this.runHintGroup
```

Comment:
Hmm, just wondering, can you write this type of clipPath/defs as part of the SVG in the Angular template? I think we can still modify the height dynamically using the d3 if we'd like, right?

Reply:
Yes, we could, just did that. Also put the rect there.